bootstrap/controllers/certificates_controller.go

```diff
@@ -0,0 +1,229 @@
+// +kubebuilder:rbac:groups=bootstrap.cluster.x-k8s.io,resources=ck8sconfigs/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;clusters/status;machines;machines/status,verbs=get;list;watch
+// +kubebuilder:rbac:groups=exp.cluster.x-k8s.io,resources=machinepools;machinepools/status,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=secrets;events;configmaps,verbs=get;list;watch;create;update;patch;delete
+func (r *CertificatesReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("namespace", req.Namespace, "machine", req.Name)
+
+	m := &clusterv1.Machine{}
+	if err := r.Get(ctx, req.NamespacedName, m); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	if !m.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Machine is being deleted, return early.
+		return ctrl.Result{}, nil
+	}
+
+	mAnnotations := m.GetAnnotations()
+
+	var refreshCertificates, hasExpiryDateAnnotation bool
+	_, refreshCertificates = mAnnotations[bootstrapv1.CertificatesRefreshAnnotation]
+	_, hasExpiryDateAnnotation = mAnnotations[bootstrapv1.MachineCertificatesExpiryDateAnnotation]
+	if !refreshCertificates && hasExpiryDateAnnotation {
+		// No need to refresh certificates or update expiry date, return early.
+		return ctrl.Result{}, nil
+	}
+
+	// Look up for the CK8sConfig.
+	config := &bootstrapv1.CK8sConfig{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: m.Namespace, Name: m.Spec.Bootstrap.ConfigRef.Name}, config); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Get the owner of the CK8sConfig to determine if it's a control plane or worker node.
+	configOwner, err := bsutil.GetConfigOwner(ctx, r.Client, config)
+	if err != nil {
+		log.Error(err, "Failed to get config owner")
+		return ctrl.Result{}, err
+	}
+	if configOwner == nil {
+		return ctrl.Result{}, nil
+	}
+
+	cluster, err := util.GetClusterByName(ctx, r.Client, m.GetNamespace(), m.Spec.ClusterName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	microclusterPort := config.Spec.ControlPlaneConfig.GetMicroclusterPort()
+	workload, err := r.managementCluster.GetWorkloadCluster(ctx, util.ObjectKey(cluster), microclusterPort)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	patchHelper, err := patch.NewHelper(m, r.Client)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create patch helper for machine: %w", err)
+	}
+
+	scope := &CertificatesScope{
+		Log:      log,
+		Machine:  m,
+		Config:   config,
+		Cluster:  cluster,
+		Patcher:  patchHelper,
+		Workload: workload,
+	}
+
+	if refreshCertificates {
+		if configOwner.IsControlPlaneMachine() {
+			return r.refreshControlPlaneCertificates(ctx, scope)
+		} else {
+			return ctrl.Result{}, fmt.Errorf("worker nodes are not supported yet")
```

Comment:
This would infinitely loop on a retry since an error is returned, we should log not supported and ignore the request by returning `ctrl.Result{}, nil`